src/com/gh4a/adapter/NotificationAdapter.java

```diff
@@ -0,0 +1,212 @@
+package com.gh4a.adapter;
+
+import android.content.Context;
+import android.support.annotation.Nullable;
+import android.support.v7.widget.RecyclerView;
+import android.view.LayoutInflater;
+import android.view.MenuItem;
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.ImageView;
+import android.widget.PopupMenu;
+import android.widget.TextView;
+
+import com.gh4a.R;
+import com.gh4a.loader.NotificationHolder;
+import com.gh4a.utils.StringUtils;
+import com.gh4a.utils.UiUtils;
+
+import org.eclipse.egit.github.core.Notification;
+import org.eclipse.egit.github.core.NotificationSubject;
+import org.eclipse.egit.github.core.Repository;
+
+public class NotificationAdapter extends
+        RootAdapter<NotificationHolder, NotificationAdapter.ViewHolder> {
+    private static final int VIEW_TYPE_NOTIFICATION_HEADER = RootAdapter.CUSTOM_VIEW_TYPE_START + 1;
+    private static final String SUBJECT_ISSUE = "Issue";
+    private static final String SUBJECT_PULL_REQUEST = "PullRequest";
+    private static final String SUBJECT_COMMIT = "Commit";
+
+    public interface OnNotificationActionCallback {
+        void markAsRead(NotificationHolder notificationHolder);
+
+        void unsubscribe(NotificationHolder notificationHolder);
+    }
+
+    private final int mBottomMargin;
+    private final Context mContext;
+    private final OnNotificationActionCallback mActionCallback;
+
+    public NotificationAdapter(Context context, OnNotificationActionCallback actionCallback) {
+        super(context);
+        mContext = context;
+        mActionCallback = actionCallback;
+
+        mBottomMargin = context.getResources().getDimensionPixelSize(R.dimen.card_margin);
+    }
+
+    public void markAsRead(@Nullable Repository repository, @Nullable Notification notification) {
+        NotificationHolder previousRepoItem = null;
+        int notificationsInSameRepoCount = 0;
+
+        boolean isMarkingSingleNotification = repository == null && notification != null;
+
+        for (int i = 0; i < getCount(); i++) {
+            NotificationHolder item = getItem(i);
+
+            // Passing both repository and notification as null will mark everything as read
+            if ((repository == null && notification == null) ||
+                    (repository != null && item.repository.equals(repository)) ||
+                    (item.notification != null && item.notification.equals(notification))) {
+                item.setIsRead(true);
+            }
+
+            // When marking single notification as read also mark the repository if it contained
+            // only 1 notification
+            if (isMarkingSingleNotification) {
+                if (item.notification == null) {
+                    if (previousRepoItem != null && notificationsInSameRepoCount == 1) {
```

Comment:
Bug: This should be run only if the repository matches notification. Same for the final check below loop.